readthedocs/core/unresolver.py

```diff
@@ -1,99 +1,279 @@
+        If the translation exists, we return a result even if the version doesn't,
+        so the translation is taken as the canonical project (useful for 404 pages).
+        """
+        match = self.multiversion_pattern.match(path)
+        if not match:
+            return None
+
+        language = match.group("language")
+        version_slug = match.group("version")
+        file = self._normalize_filename(match.group("file"))
+
+        if canonical_project.language == language:
+            project = canonical_project
+        else:
+            project = canonical_project.translations.filter(language=language).first()
 
-        None is returned if the request isn't valid.
+        if project:
+            version = project.versions.filter(slug=version_slug).first()
+            if version:
+                return project, version, file
+            return project, None, None
+
+        return None
+
+    def _match_subproject(self, canonical_project, path):
         """
-        parsed = urlparse(path)
-        path = parsed.path
-        project_slug = getattr(request, 'host_project_slug', None)
+        Try to match a subproject.
 
-        if not project_slug:
+        If the subproject exists, we try to resolve the rest of the path
+        with the subproject as the canonical project.
+        """
+        match = self.subproject_pattern.match(path)
+        if not match:
             return None
 
-        _, __, kwargs = url_resolve(
-            path,
-            urlconf='readthedocs.proxito.urls',
+        project_slug = match.group("project")
+        file = self._normalize_filename(match.group("file"))
+        project_relationship = (
+            canonical_project.subprojects.filter(alias=project_slug)
+            .prefetch_related("child")
+            .first()
         )
+        if project_relationship:
+            return self._unresolve_path(
+                canonical_project=project_relationship.child,
+                path=file,
+                check_subprojects=False,
+            )
+        return None
 
-        mixin = ServeDocsMixin()
-        version_slug = mixin.get_version_from_host(request, kwargs.get('version_slug'))
+    def _match_single_version_project(self, canonical_project, path):
+        """
+        Try to match a single version project.
 
-        final_project, lang_slug, version_slug, filename = _get_project_data_from_request(  # noqa
-            request,
-            project_slug=project_slug,
-            subproject_slug=kwargs.get('subproject_slug'),
-            lang_slug=kwargs.get('lang_slug'),
-            version_slug=version_slug,
-            filename=kwargs.get('filename', ''),
-        )
+        By default any path will match.
+        """
+        file = self._normalize_filename(path)
+        version = canonical_project.versions.filter(
+            slug=canonical_project.default_version
+        ).first()
+        if version:
```

Comment:
In what case will this be `False`? I think when they have a default_version that doesn't exist?

Reply:
yeah, this is when a use doesn't have a default version or the default version is invalid